src/advisories/cfg.rs

```diff
@@ -66,45 +67,101 @@ impl Default for Config {
 impl crate::cfg::UnvalidatedConfig for Config {
     type ValidCfg = ValidConfig;
 
-    fn validate(self, cfg_file: FileId) -> Result<Self::ValidCfg, Vec<Diagnostic>> {
+    fn validate(self, cfg_file: FileId, diags: &mut Vec<Diagnostic>) -> Self::ValidCfg {
         let mut ignored: Vec<_> = self.ignore.into_iter().map(AdvisoryId::from).collect();
         ignored.sort();
 
-        let mut db_urls = self.db_urls;
+        let mut db_urls: Vec<_> = self
+            .db_urls
+            .into_iter()
+            .filter_map(|dburl| match crate::cfg::parse_url(cfg_file, dburl) {
+                Ok(u) => Some(u),
+                Err(diag) => {
+                    diags.push(diag);
+                    None
+                }
+            })
+            .collect();
+
         if let Some(db_url) = self.db_url {
-            log::warn!("the 'db_url' option is deprecated, use 'db_urls' instead");
-            db_urls.push(db_url);
+            diags.push(
+                Diagnostic::warning()
+                    .with_message("'db_url' is deprecated, use 'db_urls' instead")
+                    .with_labels(vec![Label::primary(cfg_file, db_url.span.clone())]),
+            );
+
+            match crate::cfg::parse_url(cfg_file, db_url) {
+                Ok(url) => db_urls.push(url),
+                Err(diag) => {
+                    diags.push(diag);
+                }
+            }
+        }
+
+        db_urls.sort();
+
+        // Warn about duplicates before removing them so the user can cleanup their config
+        if db_urls.len() > 1 {
+            for window in db_urls.windows(2) {
```

Comment:
If you want to check for duplicates in the `Vec`, this will not work. This only works if the duplicate URLs if they are side-by-side in the list. E.g. In the list `[1, 2, 3, 4, 1]`, `1` will not be reported as a duplicate.
I think you have to use a nested loop here to report every duplicate, but be careful to not report an entry multiple times.

Reply:
The vector is sorted a couple of lines up.

Reply:
Oh I see😅